fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentPathValidator.java

```diff
@@ -0,0 +1,269 @@
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.fcrepo.kernel.api.exception.ExternalMessageBodyException;
+import org.slf4j.Logger;
+
+/**
+ * Validates external content paths to ensure that they are within a configured allowed list of paths.
+ *
+ * @author bbpennel
+ */
+public class ExternalContentPathValidator {
+
+    private static final Logger LOGGER = getLogger(ExternalContentPathValidator.class);
+
+    private final Set<String> ALLOWED_SCHEMES = new HashSet<>(Arrays.asList("file", "http", "https"));
+
+    private final Pattern SCHEME_PATTERN = Pattern.compile("^(http|https|file):/.*");
+
+    private final Pattern RELATIVE_MOD_PATTERN = Pattern.compile(".*(^|/)\\.\\.($|/).*");
+
+    private final Pattern NORMALIZE_FILE_URI = Pattern.compile("^file:/{2,3}");
+
+    private String configPath;
+
+    private List<String> allowedList;
+
+    private boolean monitorForChanges;
+
+    private Thread monitorThread;
+
+    private boolean monitorRunning;
+
+    /**
+     * Validates that an external path is valid. The path must be an HTTP or file URI within the allow list of paths,
+     * be absolute, and contain no relative modifier.
+     *
+     * @param extPath
+     * @throws ExternalMessageBodyException
+     */
+    public void validate(final String extPath) throws ExternalMessageBodyException {
+        if (allowedList == null || allowedList.size() == 0) {
+            throw new ExternalMessageBodyException("External content is disallowed by the server");
+        }
+
+        if (isEmpty(extPath)) {
+            throw new ExternalMessageBodyException("External content path was empty");
+        }
+
+        final String path = normalizePath(extPath.toLowerCase());
+        if (RELATIVE_MOD_PATTERN.matcher(path).matches()) {
+            throw new ExternalMessageBodyException("Path was not absolute: " + extPath);
+        }
+
+        final URI uri;
+        try {
+            uri = new URI(path);
+        } catch (final URISyntaxException e) {
+            throw new ExternalMessageBodyException("Path was not a valid URI: " + extPath);
+        }
+        if (!uri.isAbsolute()) {
+            throw new ExternalMessageBodyException("Path was not absolute: " + extPath);
+        }
+        if (!ALLOWED_SCHEMES.contains(uri.getScheme())) {
+            throw new ExternalMessageBodyException("Path did not provide an accept scheme: " + extPath);
+        }
+
+        if (allowedList.stream().anyMatch(allowed -> path.startsWith(allowed))) {
```

Comment:
`path.startsWith()` may be too permissive.
With the following rule:
```
file:///tmp/awoods/test
```
The following external content can be created:
```
file:///tmp/awoods/testsuite-execution-report.html
```

Reply:
Is this an issue? when setting up the configuration you could put `file:///tmp/awoods/test/` to avoid that. Using a simple starts with allows for specific files to be allowed without extra characters or behaviors.